File: ios/Maturita/Data/ProgressStore.swift
import Foundation

final class ProgressStore {
    private let defaults = UserDefaults.standard

    var themeId: ThemeId {
        get { ThemeId(rawValue: defaults.integer(forKey: "theme")) ?? .catppuccin }
        set { defaults.set(newValue.rawValue, forKey: "theme") }
    }

    var mode: ColorMode {
        get { defaults.string(forKey: "mode") == "dark" ? .dark : .light }
        set { defaults.set(newValue == .light ? "light" : "dark", forKey: "mode") }
    }

    var lang: UiLang {
        get { defaults.string(forKey: "lang") == "en" ? .en : .cs }
        set { defaults.set(newValue == .en ? "en" : "cs", forKey: "lang") }
    }

    var seenCommit: String {
        get { defaults.string(forKey: "seen_commit") ?? "" }
        set { defaults.set(newValue, forKey: "seen_commit") }
    }

    func germanDone(_ unit: Int, _ ex: Int) -> Bool { defaults.bool(forKey: "g.\(unit).\(ex)") }
    func markGerman(_ unit: Int, _ ex: Int) { defaults.set(true, forKey: "g.\(unit).\(ex)") }
    func vocabDone(_ unit: Int) -> Bool { defaults.bool(forKey: "g.\(unit).vocab") }
    func markVocab(_ unit: Int) { defaults.set(true, forKey: "g.\(unit).vocab") }

    func netDone(_ id: Int) -> Bool { defaults.bool(forKey: "net.\(id)") }
    func markNet(_ id: Int) { defaults.set(true, forKey: "net.\(id)") }

    func hwDone(_ id: Int) -> Bool { defaults.bool(forKey: "hw.\(id)") }
    func markHw(_ id: Int) { defaults.set(true, forKey: "hw.\(id)") }

    func mluvDone(_ n: Int) -> Bool { defaults.bool(forKey: "mluv.\(n)") }
    func markMluv(_ n: Int) { defaults.set(true, forKey: "mluv.\(n)") }

    func bookQuiz(_ id: String) -> Bool { defaults.bool(forKey: "book.\(id).quiz") }
    func markBookQuiz(_ id: String) { defaults.set(true, forKey: "book.\(id).quiz") }
    func bookPlot(_ id: String) -> Bool { defaults.bool(forKey: "book.\(id).plot") }
    func markBookPlot(_ id: String) { defaults.set(true, forKey: "book.\(id).plot") }
}

struct ProgressSum {
    var doneEx = 0
    var totalEx = 0
    var doneUnits = 0
    var openUnits = 0
}

func summarize(_ content: Content, _ p: ProgressStore) -> ProgressSum {
    var sum = ProgressSum()
    for u in content.german where u.bool("unlocked") {
        let id = u.int("id")
        let names = u.strs("names")
        sum.totalEx += names.count
        sum.openUnits += 1
        let done = names.indices.filter { p.germanDone(id, $0 + 1) }.count
        sum.doneEx += done
        if !names.isEmpty && done == names.count { sum.doneUnits += 1 }
    }
    for l in content.netLessons {
        sum.totalEx += 1
        sum.openUnits += 1
        if p.netDone(l.int("id")) {
            sum.doneEx += 1
            sum.doneUnits += 1
        }
    }
    for l in content.hw {
        sum.totalEx += 1
        sum.openUnits += 1
        if p.hwDone(l.int("id")) {
            sum.doneEx += 1
            sum.doneUnits += 1
        }
    }
    sum.totalEx += content.mluvnice.count
    sum.openUnits += 1
    let md = content.mluvnice.filter { p.mluvDone($0.int("id")) }.count
    sum.doneEx += md
    if md == content.mluvnice.count && !content.mluvnice.isEmpty { sum.doneUnits += 1 }
    return sum
}

File: android/app/src/main/java/org/maturita/maturita/data/ProgressStore.kt
package org.maturita.maturita.data

import android.content.Context

class ProgressStore(context: Context) {
    private val prefs = context.getSharedPreferences("maturita", Context.MODE_PRIVATE)

    var themeId: ThemeId
        get() = ThemeId.entries.getOrElse(prefs.getInt("theme", 0)) { ThemeId.Catppuccin }
        set(v) { prefs.edit().putInt("theme", v.ordinal).apply() }

    var mode: ColorMode
        get() = if (prefs.getString("mode", "light") == "dark") ColorMode.Dark else ColorMode.Light
        set(v) { prefs.edit().putString("mode", if (v == ColorMode.Light) "light" else "dark").apply() }

    var lang: UiLang
        get() = if (prefs.getString("lang", "cs") == "en") UiLang.En else UiLang.Cs
        set(v) { prefs.edit().putString("lang", if (v == UiLang.En) "en" else "cs").apply() }

    var seenCommit: String
        get() = prefs.getString("seen_commit", "") ?: ""
        set(v) { prefs.edit().putString("seen_commit", v).apply() }

    fun germanDone(unit: Int, ex: Int) = prefs.getBoolean("g.$unit.$ex", false)
    fun markGerman(unit: Int, ex: Int) { prefs.edit().putBoolean("g.$unit.$ex", true).apply() }
    fun vocabDone(unit: Int) = prefs.getBoolean("g.$unit.vocab", false)
    fun markVocab(unit: Int) { prefs.edit().putBoolean("g.$unit.vocab", true).apply() }

    fun netDone(id: Int) = prefs.getBoolean("net.$id", false)
    fun markNet(id: Int) { prefs.edit().putBoolean("net.$id", true).apply() }

    fun hwDone(id: Int) = prefs.getBoolean("hw.$id", false)
    fun markHw(id: Int) { prefs.edit().putBoolean("hw.$id", true).apply() }

    fun mluvDone(n: Int) = prefs.getBoolean("mluv.$n", false)
    fun markMluv(n: Int) { prefs.edit().putBoolean("mluv.$n", true).apply() }

    fun bookQuiz(id: String) = prefs.getBoolean("book.$id.quiz", false)
    fun markBookQuiz(id: String) { prefs.edit().putBoolean("book.$id.quiz", true).apply() }
    fun bookPlot(id: String) = prefs.getBoolean("book.$id.plot", false)
    fun markBookPlot(id: String) { prefs.edit().putBoolean("book.$id.plot", true).apply() }
}

data class ProgressSum(var doneEx: Int = 0, var totalEx: Int = 0, var doneUnits: Int = 0, var openUnits: Int = 0)

fun summarize(content: Content, p: ProgressStore): ProgressSum {
    val sum = ProgressSum()
    content.german.filter { it.bool("unlocked") }.forEach { u ->
        val id = u.int("id")
        val names = u.strs("names")
        sum.totalEx += names.size
        sum.openUnits += 1
        val done = names.indices.count { p.germanDone(id, it + 1) }
        sum.doneEx += done
        if (names.isNotEmpty() && done == names.size) sum.doneUnits += 1
    }
    content.netLessons.forEach { l ->
        sum.totalEx += 1
        sum.openUnits += 1
        if (p.netDone(l.int("id"))) {
            sum.doneEx += 1
            sum.doneUnits += 1
        }
    }
    content.hw.forEach { l ->
        sum.totalEx += 1
        sum.openUnits += 1
        if (p.hwDone(l.int("id"))) {
            sum.doneEx += 1
            sum.doneUnits += 1
        }
    }
    sum.totalEx += content.mluvnice.size
    sum.openUnits += 1
    val md = content.mluvnice.count { p.mluvDone(it.int("id")) }
    sum.doneEx += md
    if (md == content.mluvnice.size && content.mluvnice.isNotEmpty()) sum.doneUnits += 1
    return sum
}
